pkg/gadgettracermanager/gadgettracermanager.go

```diff
@@ -366,13 +366,28 @@ func (g *GadgetTracerManager) run() {
 				}
 			}
 		case c := <-g.createdChan:
-			containers := g.k8sClient.PodToContainers(c)
 			key, _ := cache.MetaNamespaceKeyFunc(c)
 			containerIDs, ok := containerIDsByKey[key]
 			if !ok {
 				containerIDs = make(map[string]struct{})
 				containerIDsByKey[key] = containerIDs
 			}
+
+			// first: remove containers that are not running anymore
+			nonrunning := g.k8sClient.GetNonRunningContainers(c)
+			for _, id := range nonrunning {
+				// container not found, do nothing
+				if _, ok := containerIDs[id]; !ok {
+					continue
+				}
+
+				g.RemoveContainer(nil, &pb.ContainerDefinition{
+					Id: id,
+				})
+			}
+
+			// second: add containers that are in running state
+			containers := g.k8sClient.PodToContainers(c)
```

Comment:
Isn't this variable `c` containing all containers, including the ones that are not running anymore? It might work because containers are first running before going into an error state, so `containerIDs` contains the container id, added line 399, but the code is confusing...

I suggest checking `nonrunning` just before `g.AddContainer` below so that containers created directly in the error state are not added in the first place. I think this could happen if the gadget pod is started after other pods are already in the error state. This check would be easier to implement if `GetNonRunningContainers` returns a `map[string]struct{}` instead of a `[]string`.

Reply:
`PodToContainers()` only returns containers in running state. I agree that this code is not clean at all but I don't want to touch it until we merge https://github.com/kinvolk/inspektor-gadget/pull/313, after it we'll be able to so some more cleanup on this code. I propose to merge this as it's and rework the code after https://github.com/kinvolk/inspektor-gadget/pull/313 lands, what do you think?

Reply:
Ok for me.